go/libraries/doltcore/sqle/binlogreplication/binlog_json_serialization_test.go

```diff
@@ -0,0 +1,272 @@
+// Copyright 2024 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package binlogreplication
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	gmstypes "github.com/dolthub/go-mysql-server/sql/types"
+
+	"github.com/dolthub/vitess/go/mysql"
+	"github.com/stretchr/testify/require"
+)
+
+// TestJsonSerialization_EncodedBytes tests that we can properly encode JSON data to MySQL's internal encoding.
+func TestJsonSerialization_EncodedBytes(t *testing.T) {
+	tests := []struct {
+		name     string
+		json     string
+		expected []byte
+	}{
+		// Literals
+		{
+			json:     "true",
+			expected: []byte{0x4, 0x1},
+		},
+		{
+			json:     "false",
+			expected: []byte{0x4, 0x2},
+		},
+		{
+			json:     "null",
+			expected: []byte{0x4, 0x0},
+		},
+
+		// Scalars
+		{
+			json:     `"foo"`,
+			expected: []byte{0xc, 0x3, 0x66, 0x6f, 0x6f},
+		},
+		{
+			json:     "1.0",
+			expected: []byte{0xb, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xf0, 0x3f},
+		},
+		{
+			// String values up to 127 bytes use a one byte length encoding
+			name:     "127 byte string value",
+			json:     fmt.Sprintf("%q", generateLargeString(127)),
+			expected: append([]byte{0xc, 0x7f}, []byte(generateLargeString(127))...),
+		},
+		{
+			// String values over 127 bytes use a two byte length encoding
+			name:     "128 byte string value",
+			json:     fmt.Sprintf("%q", generateLargeString(128)),
+			expected: append([]byte{0xc, 0x80, 0x1}, []byte(generateLargeString(128))...),
+		},
+		{
+			// String values up to 16,383 bytes use a two byte length encoding
+			name:     "16,383 byte string value",
+			json:     fmt.Sprintf("%q", generateLargeString(16383)),
+			expected: append([]byte{0xc, 0xff, 0x7f}, []byte(generateLargeString(16383))...),
+		},
+		{
+			// String values over 16,383 bytes use a three byte length encoding
+			name:     "16,384 byte string value",
+			json:     fmt.Sprintf("%q", generateLargeString(16384)),
+			expected: append([]byte{0xc, 0x80, 0x80, 0x1}, []byte(generateLargeString(16384))...),
+		},
```

Comment:
Do we need tests here for small/large objects/arrays?

Reply:
Yeah, that's a good suggestion. We've got 'em covered in the Vitess round trip tests below, but it's worth grabbing the bytes and adding a quick sanity check test for each here, too.